src/sentry/api/authentication.py

```diff
@@ -417,6 +422,47 @@ def authenticate_token(self, request: Request, token_str: str) -> tuple[Any, Any
         if application_is_inactive:
             raise AuthenticationFailed("UserApplication inactive or deleted")
 
+        if token.organization_id:
+            # We need to make sure the organization to which the token has access is the same as the one in the URL
+            organization = None
+            organization_context = organization_service.get_organization_by_id(
+                id=token.organization_id
+            )
+            if organization_context:
+                organization = organization_context.organization
+
+            if organization:
+                resolved_url = resolve(request.path_info)
+                target_org_id_or_slug = resolved_url.kwargs.get("organization_id_or_slug")
+                if target_org_id_or_slug:
+                    if (
+                        organization.slug != target_org_id_or_slug
+                        and organization.id != target_org_id_or_slug
```

Comment:
just to confirm, we intend to not go down this path if `organization.slug != target_slug` but `organization.id == target_id`?

i think it makes sense since the org_id is immutable but the slug can be in the process of being changes

Reply:
`target_org_id_or_slug` is what comes from API path parameter that you added, we don't know if it's id or slug. So here I'm comparing it to both id and slug and saying if it matches any of them don't throw exception.
If it doesn't match any of them throw exception.

Reply:
makes sense, thanks!

Reply:
basically yes, if the slug is changed and the partner is calling APIs using slugs it's on them, but we throw exception.